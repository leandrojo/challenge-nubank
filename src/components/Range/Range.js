import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const Wrapper = styled.div`
  position: relative;
  height: 50px;
`;

const Input = styled.input.attrs(() => ({ type: 'range' }))`
  appearance: none;
  background-color: transparent;
  box-sizing: content-box;
  height: 14px;
  margin: 18px 0;
  outline: none;
  padding: 0;
  width: 100%;

  &:focus {
    outline: none;
  }

  /* Chrome Shadow Elements */

  &::-webkit-slider-runnable-track {
    animate: 0.2s;
    background: ${p => p.theme.components.range.emptyColor};
    border-radius: 20px;
    border: none;
    cursor: pointer;
    height: 4px;
    transition: .2s ease-in-out;
    width: 100%;
  }

  &::-webkit-slider-thumb {
    appearance: none;
    background: ${p => p.theme.colors.primary};
    border-radius: 50px;
    cursor: pointer;
    height: 14px;
    margin: -5px 0 0 0;
    transition: .2s cubic-bezier(0.175, 0.885, 0.32, 1.275);
    width: 14px;
  }

  &:focus::-webkit-slider-thumb {
    transition: .2s cubic-bezier(0.175, 0.885, 0.32, 1.275);
    transform: scale(1.6);
  }

  /* Firefox Shadow Elements */

  &::-moz-range-track {
    animate: 0.2s;
    background: ${p => p.theme.components.range.emptyColor};
    border-radius: 20px;
    border: none;
    cursor: pointer;
    height: 4px;
    transition: .2s ease-in-out;
    width: 100%;
  }

  &::-moz-range-thumb {
    appearance: none;
    background: ${p => p.theme.colors.primary};
    border: none;
    border-radius: 50px;
    cursor: pointer;
    height: 14px;
    margin-top: -5px;
    transition: .2s cubic-bezier(0.175, 0.885, 0.32, 1.275);
    transform: translate3d(0, 0, 10px);
    width: 14px;
  }

  &:focus::-moz-range-thumb {
    transition: .2s cubic-bezier(0.175, 0.885, 0.32, 1.275);
    transform: scale(1.6);
  }

  &::-moz-focus-outer {
    border: 0;
  }
`;

const Fill = styled.div`
  background-color: ${p => p.theme.colors.primary};
  border-radius: 20px 0 0 20px;
  height: 4px;
  margin: 0;
  padding: 0;
  pointer-events: none;
  position: absolute;
  top: calc(50% - 2px);

  -moz-top: calc(50% - 2px);
`;

export const Range = ({
  max,
  min,
  name,
  onChange,
  value,
}) => {
  const handleChange = (ev) => {
    const newValue = ev.target.value;
    onChange(newValue, name);
  };

  const percentage = ((value - min) * 100) / (max - min);

  return (
    <Wrapper>
      <Input
        max={max}
        min={min}
        name={name}
        onChange={handleChange}
        value={value}
      />
      <Fill style={{ width: `${percentage}%` }} />
    </Wrapper>
  );
};

Range.defaultProps = {
  max: 1,
  min: 0,
  name: '',
  onChange: () => {},
  value: 0,
};

Range.propTypes = {
  max: PropTypes.number,
  min: PropTypes.number,
  name: PropTypes.string,
  onChange: PropTypes.func,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

export default Range;
