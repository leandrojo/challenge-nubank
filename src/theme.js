const fontFamily = 'Rubik, sans-serif';

const theme = {
  backgroundColor: '#f5f5f5',
  colors: {
    primary: '#7f39b7',
    grayExtraDark: '#555',
    grayDark: '#767676',
    gray: '#999',
    grayLight: '#ccc',
    grayExtraLight: '#FAFAFA',
  },
  fontFamily,
  fontSize: '1em',
  typography: {
    heading: {
      color: '#252226',
    },
    paragraph: {
      color: '#716d73',
    },
  },
};

theme.components = {
  input: {
    css: {
      backgroundColor: 'transparent',
      fontFamily,
      fontSize: '1.4em',
      fontWeight: 300,
    },
  },
  pane: {
    backgroundColor: '#fff',
  },
  range: {
    emptyColor: theme.colors.grayLight,
  },
};

export default theme;
