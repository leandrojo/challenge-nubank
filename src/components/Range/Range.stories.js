/* eslint-disable import/no-extraneous-dependencies */

import React, { useState } from 'react';

import Range from './Range';

export default {
  title: 'components/Range',
};

export const shortInterval = () => {
  const [value, setValue] = useState(0);

  return (
    <>
      <Range
        min={1}
        max={5}
        onChange={setValue}
        value={value}
      />
      <p>{`Min: 1 | Max: 5 | Value: ${value}`}</p>
    </>
  );
};

shortInterval.story = {
  name: 'short interval',
};

export const largeInterval = () => {
  const [value, setValue] = useState(0);

  return (
    <>
      <Range
        min={1}
        max={1000}
        onChange={setValue}
        value={value}
      />
      <p>{`Min: 1 | Max: 1000 | Value: ${value}`}</p>
    </>
  );
};

largeInterval.story = {
  name: 'large interval',
};

export const initialValue = () => {
  const [value, setValue] = useState(10);

  return (
    <>
      <Range
        min={1}
        max={20}
        onChange={setValue}
        value={value}
      />
      <p>{`Min: 1 | Max: 20 | Initial Value: 10 | Value: ${value}`}</p>
    </>
  );
};

initialValue.story = {
  name: 'with initial value',
};
