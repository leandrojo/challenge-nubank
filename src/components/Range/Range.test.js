/* eslint-env jest */

import React, { useState } from 'react';
import { ThemeProvider } from 'styled-components';
import renderer from 'react-test-renderer';

import theme from 'theme';

import Range from './Range';

describe('Range', () => {
  it('with initial value', () => {
    const Component = () => {
      const [value, onChange] = useState(500);

      return (
        <ThemeProvider theme={theme}>
          <Range max={1000} min={1} name="period" onChange={onChange} value={value} />
        </ThemeProvider>
      );
    };

    const expected = renderer.create(<Component />).toJSON();

    expect(expected).toMatchSnapshot();
  });
});
