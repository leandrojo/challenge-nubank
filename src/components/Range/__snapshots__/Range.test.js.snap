// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range with initial value 1`] = `
<div
  className="sc-bdVaJa QxiRk"
>
  <input
    className="sc-bwzfXH iORXor"
    max={1000}
    min={1}
    name="period"
    onChange={[Function]}
    type="range"
    value={500}
  />
  <div
    className="sc-htpNat dkFsgT"
    style={
      Object {
        "width": "49.949949949949946%",
      }
    }
  />
</div>
`;
